src/lib.rs

```diff
@@ -1,83 +1,157 @@
+//!   parsing. Right now, only [`Error::DoesNotRequireValue`] can be
+//!   returned by [`Options::next`] - but this is an implementation
+//!   detail.
+//!
+//! - The inner [`Option`] is because when a positional argument is
+//!   encountered (a non-option that isn't itself a value for one),
+//!   there are no more options to parse.
+//!
+//! Once you have an [`Opt`], you can match on it, and do one of the
+//! following things:
+//!
+//! - Call [`Options::value`], which will tell `getargs` that this
+//!   option requires a value, and either return it or raise an
+//!   [`Error::RequiresValue`].
+//!
+//! - Call [`Options::value_opt`], which will tell `getargs` that this
+//!   option may have an *optional* value, and return the value, if any.
+//!
+//! - Do nothing before the next call to [`Options::next`], which will
+//!   by default assume that the option should not have had a value (and
+//!   raise an [`Error::DoesNotRequireValue`] if an explicit value was
+//!   found).
+//!
+//! You can continue doing this until `next` returns `None`. Once it
+//! does, you can call [`Options::arg`] to get positional arguments one
+//! by one, [`Options::args`] to get an iterator over positional
+//! arguments, or [`Options::into_args`] to turn the [`Options`] into
+//! an iterator over the rest of the positional arguments.
+//!
+//! This pattern lends itself quite well to [`while let`][while let]
+//! loops:
+//!
+//! ```
+//! # use getargs::{Options, Opt};
+//! # let iter = ["-h", "--help", "-v", "--version"].into_iter();
+//! let mut opts = Options::new(iter);
+//!
+//! while let Some(opt) = opts.next()? {
+//!     match opt {
+//!         Opt::Short('h') | Opt::Long("help") => { /* ... */ }
+//!         Opt::Short('v') | Opt::Long("version") => { /* ... */ }
+//!         _ => panic!("Unknown option: {}", opt)
+//!     }
 //! }
+//! # Ok::<(), getargs::Error<&'static str>>(())
 //! ```
+//!
+//! However, [`Options`] notably does not implement [`Iterator`] because
+//! a `for` loop borrows the iterator for the duration of the loop,
+//! which would make it impossible to call [`Options::value`] and
+//! friends.
+//!
+//! # Examples
+//!
+//! Here is the full source code of an example (`print`) that prints all
+//! the flags and positional arguments it is given:
+//!
+#![doc = "```"]
+#![doc = include_str!("../examples/print.rs")]
+#![doc = "```"]
+//!
+//! First it collects all process arguments as [`String`]s (panicking if
+//! they are not valid UTF-8, as per [`args`][std::env::args]), then it
+//! creates an [`Options`] from an iterator over the arguments, then it
+//! uses a [while let][while let] loop to call [`Options::next`]. Each
+//! iteration, it consumes whatever value the option requires (if any).
+//!
+//! After there are no more options left, it uses [`Options::args`] to
+//! iterate over the rest of the arguments as positional.
+//!
+//! Here is an example that accepts options after positional arguments
```

Comment:
I think we can just put 1 example here in the crate documentation. In the `examples/` dir we could put a README.md describing each example file, and then link to that here.

Reply:
Other READMEs aren't visible from rustdoc. I think they're included in `src` but that would be weird to link to.

Reply:
I mean just link to the README file on GitHub. e.g.

```markdown
[see more examples here](https://github.com/j-tai/getargs/tree/master/examples)
```

Reply:
I went and did it anyway - linking to GitHub is suboptimal, but it's all we got I suppose.